File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/country/CountryViewModel.java
package com.dewonderstruck.apps.ashx0.viewmodel.country;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.dewonderstruck.apps.Config;
import com.dewonderstruck.apps.ashx0.repository.country.CountryRepository;
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData;
import com.dewonderstruck.apps.ashx0.utils.Utils;
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel;
import com.dewonderstruck.apps.ashx0.viewobject.Country;
import com.dewonderstruck.apps.ashx0.viewobject.common.Resource;

import java.util.List;

import javax.inject.Inject;

public class CountryViewModel extends PSViewModel {
    private LiveData<Resource<List<Country>>> countryListData;
    private MutableLiveData<CountryViewModel.TmpDataHolder> countryListObj = new MutableLiveData<>();

    private LiveData<Resource<Boolean>> nextPageCountryListData;
    private MutableLiveData<CountryViewModel.TmpDataHolder> nextPageCountryListObj = new MutableLiveData<>();

    @Inject
    CountryViewModel(CountryRepository repository) {
        Utils.psLog("Inside CountryViewModel");

        countryListData = Transformations.switchMap(countryListObj, obj -> {

            if (obj == null) {
                return AbsentLiveData.create();
            }

            return repository.getCountryListWithShopId(Config.API_KEY, obj.shopId, obj.limit, obj.offset);
        });

        nextPageCountryListData = Transformations.switchMap(nextPageCountryListObj, obj -> {

            if (obj == null) {
                return AbsentLiveData.create();
            }

            return repository.getNextPageCityListWithShopId(Config.API_KEY, obj.shopId, obj.limit, obj.offset);
        });
    }

    public void setCountryListObj(String shopId, String limit, String offset) {
        CountryViewModel.TmpDataHolder tmpDataHolder = new CountryViewModel.TmpDataHolder();
        tmpDataHolder.shopId = shopId;
        tmpDataHolder.limit = limit;
        tmpDataHolder.offset = offset;

        countryListObj.setValue(tmpDataHolder);
    }

    public LiveData<Resource<List<Country>>> getCountryListData() {

        return countryListData;
    }

    public void setNextPageCountryListObj(String shopId, String limit, String offset) {
        CountryViewModel.TmpDataHolder tmpDataHolder = new CountryViewModel.TmpDataHolder();
        tmpDataHolder.shopId = shopId;
        tmpDataHolder.limit = limit;
        tmpDataHolder.offset = offset;

        nextPageCountryListObj.setValue(tmpDataHolder);
    }

    public LiveData<Resource<Boolean>> getNextPageCountryListData() {

        return nextPageCountryListData;
    }

    class TmpDataHolder {
        public String loginUserId = "";
        public String limit = "";
        public String offset = "";
        public String shopId = "";

    }
}

File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/country/CountryViewModel.kt
package com.dewonderstruck.apps.ashx0.viewmodel.country

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.dewonderstruck.apps.Config
import com.dewonderstruck.apps.ashx0.repository.country.CountryRepository
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData
import com.dewonderstruck.apps.ashx0.utils.Utils
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel
import com.dewonderstruck.apps.ashx0.viewobject.Country
import com.dewonderstruck.apps.ashx0.viewobject.common.Resource
import javax.inject.Inject

class CountryViewModel @Inject internal constructor(repository: CountryRepository) : PSViewModel() {
    val countryListData: LiveData<Resource<List<Country>>>
    private val countryListObj = MutableLiveData<TmpDataHolder>()
    val nextPageCountryListData: LiveData<Resource<Boolean>>
    private val nextPageCountryListObj = MutableLiveData<TmpDataHolder>()
    fun setCountryListObj(shopId: String, limit: String, offset: String) {
        val tmpDataHolder = TmpDataHolder()
        tmpDataHolder.shopId = shopId
        tmpDataHolder.limit = limit
        tmpDataHolder.offset = offset
        countryListObj.value = tmpDataHolder
    }

    fun setNextPageCountryListObj(shopId: String, limit: String, offset: String) {
        val tmpDataHolder = TmpDataHolder()
        tmpDataHolder.shopId = shopId
        tmpDataHolder.limit = limit
        tmpDataHolder.offset = offset
        nextPageCountryListObj.value = tmpDataHolder
    }

    internal inner class TmpDataHolder {
        var loginUserId = ""
        var limit = ""
        var offset = ""
        var shopId = ""
    }

    init {
        Utils.psLog("Inside CountryViewModel")
        countryListData = Transformations.switchMap(countryListObj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<Resource<List<Country>>>()
            }
            repository.getCountryListWithShopId(Config.API_KEY, obj.shopId, obj.limit, obj.offset)
        }
        nextPageCountryListData = Transformations.switchMap(nextPageCountryListObj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<Resource<Boolean>>()
            }
            repository.getNextPageCityListWithShopId(Config.API_KEY, obj.shopId, obj.limit, obj.offset)
        }
    }
}